### PR TITLE
Fix reading bytes message data after extracting message properties.

### PR DESCRIPTION
**Pull request by tnleeuw**

### CODE REVIEW

**Review by Yale96 — APPROVED** [review]

File: core/src/main/java/nl/nn/adapterframework/jms/BytesMessageInputStream.java [hunk 1]
```diff
@@ -32,14 +33,19 @@ public BytesMessageInputStream(BytesMessage bytesMsg) {
 	@Override
 	public int read() throws IOException {
 		try {
-			return bytesMsg.readByte();
+			byte[] data = new byte[1];
+			int len = bytesMsg.readBytes(data);
+			if (len == -1) {
+				return -1;
+			}
+			return data[0] & 0xFF;
```
Comment (jkosternl) [inline core/src/main/java/nl/nn/adapterframework/jms/BytesMessageInputStream.java:41]: Why this interesting bit flipping? Can you add a comment on the purpose of this magic?

File: core/src/main/java/nl/nn/adapterframework/jms/BytesMessageInputStream.java [hunk 2]
```diff
@@ -48,12 +54,17 @@ public int read(byte[] b) throws IOException {
...
-	public int read(byte[] b, int off, int len) throws IOException {
+	public int read(@Nonnull byte[] b, int off, int len) throws IOException {
 		try {
-			byte[] readbuf = new byte[len];
-			int result = bytesMsg.readBytes(readbuf);
+			// If we try filling the whole array from start, then we do not need to read data into
+			// a temp array and copy it.
+			if (off == 0 && len == b.length) {
+				return read(b);
+			}
```
Comment (jkosternl) [inline core/src/main/java/nl/nn/adapterframework/jms/BytesMessageInputStream.java:63]: Just to understand your fix: this if-statement at line 61 is an optimization right, not a fix?

File: core/src/main/java/nl/nn/adapterframework/extensions/esb/EsbJmsListener.java [hunk 3]
```diff
@@ -136,34 +136,37 @@ public Map<String, Object> extractMessageProperties(Message rawMessage) throws L
...
-				TextMessage textMessage = (TextMessage) rawMessage;
-				soapMessage = textMessage.getText();
-			} else if (rawMessage instanceof BytesMessage) {
-				BytesMessage bytesMessage = (BytesMessage) rawMessage;
-				InputStream input = new BytesMessageInputStream(bytesMessage);
-				soapMessage = StreamUtil.streamToString(input);
-			} else {
-				soapMessage = null;
-			}
+		if (!getxPathLogMap().isEmpty()) {
```
Comment (jkosternl) [inline core/src/main/java/nl/nn/adapterframework/extensions/esb/EsbJmsListener.java:139]:
```suggestion
		if (getxPathLogMap().isEmpty()) {
		     return messageProperties;
		}
```

**Review by jkosternl — APPROVED** [review]

File: core/src/main/java/nl/nn/adapterframework/jms/BytesMessageInputStream.java (hunk 1 above)
Comment (tnleeuw) [reply core/src/main/java/nl/nn/adapterframework/jms/BytesMessageInputStream.java:41]: This was suggested by IntelliJ to ensure the return data was unsigned. You are absolutely right this needs a comment with explanation.

File: core/src/main/java/nl/nn/adapterframework/extensions/esb/EsbJmsListener.java [hunk 4]
```diff
@@ -136,34 +136,37 @@ public Map<String, Object> extractMessageProperties(Message rawMessage) throws L
...
+				if (rawMessage instanceof TextMessage) {
+					TextMessage textMessage = (TextMessage) rawMessage;
+					soapMessage = textMessage.getText();
+				} else if (rawMessage instanceof BytesMessage) {
+					BytesMessage bytesMessage = (BytesMessage) rawMessage;
+					InputStream input = new BytesMessageInputStream(bytesMessage);
+					soapMessage = StreamUtil.streamToString(input);
+					bytesMessage.reset();
+				} else {
+					soapMessage = null;
```
Comment (Laurens-makel) [inline core/src/main/java/nl/nn/adapterframework/extensions/esb/EsbJmsListener.java:151]: Just wondering, is this explicit `null` assignment needed or is it just a code style preference?

Comment (Laurens-makel) [reply core/src/main/java/nl/nn/adapterframework/extensions/esb/EsbJmsListener.java]: I thought that a `String` is by default `null` when it's not initialized right?

Comment (tnleeuw) [reply core/src/main/java/nl/nn/adapterframework/extensions/esb/EsbJmsListener.java]: It's a local variable, not a field in the class. So when not initialized it has no value. You cannot access it unless it's explicitly assigned -- the compiler will give an error.

Comment (Laurens-makel) [reply core/src/main/java/nl/nn/adapterframework/extensions/esb/EsbJmsListener.java]: I understand, although I think it's a bit silly that local variables behave differently.

Thanks for the clarification :)

Comment (tnleeuw) [reply core/src/main/java/nl/nn/adapterframework/extensions/esb/EsbJmsListener.java]: Actually I'm glad they behave the way they do, as it allows to catch accidentally unitialized variables.

Comment (Laurens-makel) [reply core/src/main/java/nl/nn/adapterframework/extensions/esb/EsbJmsListener.java]: > Actually I'm glad they behave the way they do, as it allows to catch accidentally unitialized variables.

That's a good argument.

For me it's the inconsistency, apparently only null is assigned to class member variables.. why wouldn't the same concept apply here?

Comment (tnleeuw) [reply core/src/main/java/nl/nn/adapterframework/extensions/esb/EsbJmsListener.java]: Yes, I agree.

**Comment by jkosternl** [comment]
Looking good now, with the extra changes. Thanks 👍🏻

**Review by nielsm5 — APPROVED** [review]

**Comment by tnleeuw** [comment]
Since I'm still adding a test for IAF-Test I put the PR back into "Draft" for now, I expect to be ready with that later today.

**Comment by tnleeuw** [comment]
> Since I'm still adding a test for IAF-Test I put the PR back into "Draft" for now, I expect to be ready with that later today.

I've now added integration tests, verified that the integration test with BytesMessage indeed breaks without the applied fixes.

**Review by jkosternl — APPROVED** [review]

**Review by nielsm5 — APPROVED** [review]

**Pull request merged**